packages/parse5-htmlparser2-tree-adapter/lib/index.ts

```diff
@@ -0,0 +1,264 @@
+import {
+    Node,
+    NodeWithChildren,
+    Element,
+    Document,
+    ProcessingInstruction,
+    Comment,
+    Text,
+    isDirective,
+    isText,
+} from 'domhandler';
+export { isComment as isCommentNode, isTag as isElementNode, isText as isTextNode } from 'domhandler';
+
+export type Htmlparser2TreeAdapterMap = TreeAdapterTypeMap<
+    Node,
+    NodeWithChildren,
+    Node,
+    Document,
+    Document,
+    Element,
+    Comment,
+    Text,
+    Element,
+    ProcessingInstruction
+>;
+
+//Node construction
+export function createDocument(): Document {
+    const node = new Document([]);
+    node['x-mode'] = DOCUMENT_MODE.NO_QUIRKS;
+    return node;
+}
+
+export function createDocumentFragment(): Document {
+    return new Document([]);
+}
+
+export function createElement(tagName: string, namespaceURI: NS, attrs: Attribute[]): Element {
+    const attribs = Object.create(null);
+    const attribsNamespace = Object.create(null);
+    const attribsPrefix = Object.create(null);
+
+    for (let i = 0; i < attrs.length; i++) {
+        const attrName = attrs[i].name;
+
+        attribs[attrName] = attrs[i].value;
+        attribsNamespace[attrName] = attrs[i].namespace;
+        attribsPrefix[attrName] = attrs[i].prefix;
+    }
+
+    const node = new Element(tagName, attribs, []);
+    node.namespace = namespaceURI;
+    node['x-attribsNamespace'] = attribsNamespace;
+    node['x-attribsPrefix'] = attribsPrefix;
+    return node;
+}
+
+export function createCommentNode(data: string): Comment {
+    return new Comment(data);
+}
+
+function createTextNode(value: string): Text {
+    return new Text(value);
+}
+
+//Tree mutation
+export function appendChild(parentNode: NodeWithChildren, newNode: Node): void {
+    const prev = parentNode.children[parentNode.children.length - 1];
+
+    if (prev) {
+        prev.next = newNode;
+        newNode.prev = prev;
+    }
+
+    parentNode.children.push(newNode);
+    newNode.parent = parentNode;
+}
+
+export function insertBefore(parentNode: NodeWithChildren, newNode: Node, referenceNode: Node): void {
+    const insertionIdx = parentNode.children.indexOf(referenceNode);
+    const { prev } = referenceNode;
+
+    if (prev) {
+        prev.next = newNode;
+        newNode.prev = prev;
+    }
+
+    referenceNode.prev = newNode;
+    newNode.next = referenceNode;
+
+    parentNode.children.splice(insertionIdx, 0, newNode);
+    newNode.parent = parentNode;
+}
+
+export function setTemplateContent(templateElement: Element, contentElement: Document): void {
+    appendChild(templateElement, contentElement);
+}
+
+export function getTemplateContent(templateElement: Element): Node {
+    return templateElement.children[0];
+}
+
+export function setDocumentType(
+    document: Document,
+    name: string | null,
+    publicId: string | null,
+    systemId: string | null
+): void {
+    const data = doctype.serializeContent(name, publicId, systemId);
+    let doctypeNode = document.children.find(
```

Comment:
you can drop the cast if you use type predicate: `find((node): node is ProcessingInstruction => ...)`